### PR TITLE
As a user, I can rest easy in the knowledge that Pulp's Celery tasks are not serialized dangerously

### PR DESCRIPTION
**Pull request by ipanova**
https://pulp.plan.io/issues/23

### CODE REVIEW

**Comment by bmbouter** [comment]
I think the code is not Python 2.6 compliant. Can you update the PR so that Jenkins is happy.

**Comment by bmbouter** [comment]
I'm glad the commits are preserved. Any changes to this PR can you squash into your commit, and preserve the other 2 (as you've already done).

**Comment by ipanova** [comment]
ok test

**Comment by bmbouter** [comment]
Please add details to the commit message outlining the changes. I think the topics are:

1) the inheritance changes to introduce PulpTask
2) the serialization/deserialization layer which handles ObjectIds
3) The UnitAssociation issues to_json() from_json()
4) The changes to criteria

File: server/pulp/server/db/model/criteria.py [hunk 1]
```diff
@@ -177,6 +177,46 @@ def __init__(self, type_ids=None, association_filters=None, unit_filters=None,
...
+            'unit_fields': self.unit_fields,
+            'remove_duplicates': self.remove_duplicates
+        }
+
+    @classmethod
+    def from_dict(cls, input_dictionary):
+        """
+        Convert a dictionary representation of the UnitAssociationCriteria into a new
+        UnitAssociationCriteria object.
+        The output of to_dict() is suitable as input to this method.
```
Comment (bmbouter) [inline server/pulp/server/db/model/criteria.py]: Consider adding a newline above this to separate it from the paragraph above it.

File: server/pulp/server/db/model/criteria.py [hunk 2]
```diff
@@ -177,6 +177,46 @@ def __init__(self, type_ids=None, association_filters=None, unit_filters=None,
...
+
+        :param input_dictionary: The dictionary representation of a UnitAssociationCriteria
+                                 object that will be used to construct one.
+        :type  input_dictionary: dict
+        :return:                 A new UnitAssociationCriteria object
```
Comment (bmbouter) [inline server/pulp/server/db/model/criteria.py]: I usually add a newline above the :return to separate the inputs from the outputs.

File: server/pulp/server/controllers/distributor.py [hunk 3]
```diff
@@ -107,3 +107,35 @@ def update(repo_id, distributor_id, config, delta):
         bind_error = PulpCodedException(PLP0002, repo_id=repo_id, distributor_id=distributor_id)
         bind_error.child_exceptions = bind_errors
     return TaskResult(distributor, bind_error, additional_tasks)
+
+
+@celery.task(base=PulpTask)
+def publish(repo_id, distributor_id, overrides=None):
```
Comment (bmbouter) [inline server/pulp/server/controllers/distributor.py]: Was this moved from somewhere else? I can't think of why this is added.

File: server/pulp/server/controllers/distributor.py [hunk 4]
```diff
@@ -107,3 +107,35 @@ def update(repo_id, distributor_id, config, delta):
...
+    :param overrides: dictionary of options to pass to the publish manager
+    :type overrides: dict or None
+    :return: list of call requests
+    :rtype: list
+    """
+    return managers.repo_publish_manager().queue_publish(repo_id, distributor_id, overrides)
+
+
+@celery.task(base=PulpTask)
+def sync_with_auto_publish(repo_id, overrides=None):
```
Comment (bmbouter) [inline server/pulp/server/controllers/distributor.py]: Was this moved from somewhere else? I can't think of why this is added.

File: server/test/unit/server/async/test_tasks.py [hunk 5]
```diff
@@ -1059,3 +1059,12 @@ def test_is_not_worker_is_scheduler(self):
 
     def test_is_not_worker_is_resource_mgr(self):
         self.assertEquals(tasks._is_worker("resource_manager@some.hostname"), False)
+
+
+class TestPulpTask(unittest.TestCase):
+
+    def check_task_type(self):
```
Comment (bmbouter) [inline server/test/unit/server/async/test_tasks.py]: Can a docstring be added to this indicating that its purpose is to have all Celery tasks defined in Pulp inherit from PulpTask and not from Celery directly. Consider adding that it should not be adjusted unless you really know what you are doing. Normally I don't docstring tests, but this one is important.

File: server/test/unit/server/async/test_tasks.py [hunk 6]
```diff
@@ -1059,3 +1059,12 @@ def test_is_not_worker_is_scheduler(self):
...
+        for task_name, task in app.celery.tasks.iteritems():
+            if task_name.startswith('pulp'):
+                if not isinstance(task, tasks.PulpTask):
+                    self.fail('task named %s must inherit from %s' % (task_name, tasks.PulpTask))
```
Comment (bmbouter) [inline server/test/unit/server/async/test_tasks.py]: This needs to be adjusted slightly to be.  `'task named %s must have %s as an ancestor'`

**Comment by bmbouter** [comment]
@ipanova This looks really great, I left a bunch of small comments to handle. The two things I don't understand and I believe should be removed are these comments https://github.com/pulp/pulp/pull/2241#discussion_r47375889 and https://github.com/pulp/pulp/pull/2241#discussion_r47375896

Once all of that gets fixed I think it will be good for merging. I can review again on Monday if you want or not, or you can merge it otherwise.

**Comment by ipanova** [comment]
ok test

**Comment by ipanova** [comment]
ok test

**Comment by ipanova** [comment]
ok test

**Comment by bmbouter** [comment]
@ipanova The most recent push addresses the technical concerns. Can you rebase against the latest master so it can be cleanly merged?

**Comment by ipanova** [comment]
ok test

**Comment by ipanova** [comment]
ok test

**Comment by ipanova** [comment]
@bmbouter conflicts resolved, jenkins is happy :)

**Comment by bmbouter** [comment]
@ipanova great job! Also thanks to @shubham90 for his work on this too.

**Pull request merged**